### ISSUE TITLE
Chrome: differentiate channels

### ISSUE THREAD

**Issue by vweevers** [vweevers/win-detect-browsers#25]
I don't have a bullet-proof solution for Chrome yet. A late response to @tomitm ([#22](https://github.com/vweevers/win-detect-browsers/issues/22#issuecomment-179487666)):

> 1) I found `ChromeSSHTM.SCXTU77WPPCZZWL7LX4LSWLZ34` beside it for canary (assuming you can use a wildcard). Or is [this] key under Classes not guaranteed to be there?
> 2) I noticed that it seems to be possible to figure out if a Chrome install is stable/beta/dev/canary via the updater. The keys `HKCU\SOFTWARE\Google\Update\ClientState\{...}\ap` and `HKLM\SOFTWARE\WOW6432Node\Google\Update\ClientState\{...}\ap` have the value of `x64-dev-multi-install` or `x64-canary` for me (multi-install denotes it's installed under AppData).

1) Sadly no, this is not a universally unique id. 2) I don't see the same keys and values on my machine and I reckon the specifics of these `ClientState` keys can change between versions of chrome, or could be incomplete if the user has disabled the updater.

In general, I'm worried that differentiation of channels based on registry keys alone would increase the likelihood of false positives (there's no reliable intel, yet many edge cases).

Resources that may prove useful:
- [OmahaProxy](https://omahaproxy.appspot.com/) and an [alternative](http://10.omahaproxy-hrd.appspot.com/history)
- [node-omaha](https://github.com/DamonOehlman/node-omaha)

**Comment by tomitm** [comment]
> 1) Sadly no, this is not a universally unique id.

Is it possible to just match if the key starts with `ChromeSSHTM`? The `SS` part seems consistent with `SxS` that I've seen for Canary's installation folder (`C:\Users\tm\AppData\Local\Google\Chrome SxS`).

Of course this is assuming it's only possible to have two Chrome installations - stable/beta/dev and canary.

> In general, I'm worried that differentiation of channels based on registry keys alone would increase the likelihood of false positives (there's no reliable intel, yet many edge cases).

Yeah, I agree that this isn't the most concrete way of discovering the version, as it relies on the updater. Will keep digging.

I'm not sure how useful Omaha will be. Looking up a version number doesn't say which channel it's on. And even if it did, a Chrome installation may be outdated if it hasn't been launched in a while, so the channel might not be reliable. Depending on if Chrome promotes a given version to different channels, or a given version number is always it's original channel.

**Comment by vweevers** [comment]
> Is it possible to just match if the key starts with ChromeSSHTM? The SS part seems consistent with SxS that I've seen for Canary's installation folder (C:\Users\tm\AppData\Local\Google\Chrome SxS).

Would have to enumerate all classes. But maybe we can first read the values under `HKCR\.htm\OpenWithProgids` (as well as .html), which contains `ChromiumHTM.XXX`, `ChromeSSHTM.XXX` and `ChromeHTML` (as well as id's of other browsers). Then read `Classes\[id]\Application` to get the browser's path and name.

Combined with #24, this should be sufficient for most cases. And when it isn't, `win-detect-browsers` would still detect Chrome, but without channel info.

Could you do some research on OpenWithProgids and Classes? E.g. if there are differences between Windows versions, access restrictions, edge cases, etc..

> I'm not sure how useful Omaha will be. Looking up a version number doesn't say which channel it's on. And even if it did, a Chrome installation may be outdated if it hasn't been launched in a while, so the channel might not be reliable. Depending on if Chrome promotes a given version to different channels, or a given version number is always it's original channel.

I'm not sure either. Was thinking of comparing version and mtime of the executable with omaha's data. But even that wouldn't work, because a version is sometimes released to multiple channels at the same time.

![image](https://cloud.githubusercontent.com/assets/3055345/13740286/3c607768-e9d0-11e5-9d37-5c2e56148cf5.png)

So yeah, scratch that

**Comment by tomitm** [comment]
Sounds good, I'll see if I can dig around and figure out the details for `OpenWithProgIds` and `Classes`. Will let you know what I find in a bit.

Initial concern would be that some users might modify `OpenWithProgIds` as it seems like it's tied to the "Open With..." context menu. But, at least we have a fallback.

**Comment by tomitm** [comment]
@vweevers As long as you aren't planning on enumerating `OpenWithProgIds`, as it also contains non-browsers (editors), it looks like it should work. There's also a couple of other extensions you could use, but it seems like only Chrome registers on them.

https://msdn.microsoft.com/en-us/library/windows/desktop/cc144148(v=vs.85).aspx#fa_optional_keys_attributes

Basically `HKCR` is read-only, composed of merging `HKLM\Software\Classes` and `HKCU\Software\Classes` (which are writable), so it seems to be a good way of getting a complete list of what's available on the computer (and also means we only need to check `HKCR`). From there, it's expected implementation of a verb (*open*) is under `HKCR\Applications\{appid}\shell\{verb}` which can include `command` and `DropTarget`.

https://msdn.microsoft.com/en-us/library/windows/desktop/ee872121(v=vs.85).aspx

This article explains how `ShellExecuteEx` works and describes `HKCU\Software\Microsoft\Windows\CurrentVersion\AppPaths`, which is used for mapping a file name to a fully qualified path. Doesn't seem comprehensive to me, as I don't see all the browsers I have installed here. There's also a bit more info about `HKCR` on this page. It seems to be very similar to `\Applications`. 

There's also `HKCR\SystemFileAssociations`, but that didn't seem to have anything relevant.

https://msdn.microsoft.com/en-us/library/bb166549.aspx

> The **OpenWithProgids key is only supported in Windows XP**. Because other operating systems ignore this key, do not use it as the only registration for file handlers. Use this key to provide a better user experience in Windows XP.

Seems to be available from at least XP.

https://msdn.microsoft.com/en-us/library/windows/desktop/cc144148(v=vs.85).aspx

There's also `OpenWithList`, but Microsoft has a clear *Do not use* on that. It's [apparently](https://msdn.microsoft.com/en-us/library/windows/desktop/hh127445(v=vs.85).aspx) intended for legacy apps prior to XP.

**Comment by vweevers** [comment]
That's excellent! Thank you very much.

Correct me if I'm wrong, but if the "appid" in `HKCR\Applications\{appid}` is an executable name like `chrome.exe` - then this is pretty much useless because only one chrome channel can take that key. Or well, sofware vendors *could* use any appid they want, because it's virtual, for example, Chrome beta could use `HKCR\Applications\chrome_beta.exe` to differentiate itself. But they don't.

> The OpenWithProgids key is only supported in Windows XP.

So what these docs mean to say is "not supported before XP"? If so, we're good, and we can get by without `Applications` or `AppPaths`.

**Comment by tomitm** [comment]
> Correct me if I'm wrong, but if the "appid" in `HKCR\Applications\{appid}` is an executable name like `chrome.exe` - then this is pretty much useless because only one chrome channel can take that key.

No, it's not an executable name, fortunately. Although, I didn't seem to come across anything regarding conflict resolution in that case, now that you mention it...

Chrome stable/beta/dev use `ChromeHTML` and canary uses `ChromeSSHTM.{uniqueid}`.

It does seem to be an issue for Firefox, as I see `FirefoxHTML` under `HKCR`, which is from `HKLM\SOFTWARE\Classes\FirefoxHTML`, which points to Firefox Dev, but I can't seem to find the stable version (which was installed first) using this method. I assume stable used `FirefoxHTML` as it's key, and got overwritten during the dev install. Keeping in mind I'm pretty sure I installed both as machine (all users) instead of current user.

So that means using `HKCR\Applications` isn't a full replacement for all browers, but it's a good place to look for some.

With Firefox, it seems the best would be to use `HKLM\SOFTWARE\WOW6432Node\Mozilla\{product}\{version}\Main\PathToExe`, where `{product}` is `Mozilla Firefox` or `Firefox Developer Edition` and `{version}` is the value of `CurrentVersion` under the product key. Which is what you currently do.

> So what these docs mean to say is "not supported before XP"? 

Yeah, I'm assuming they're outdated docs, considering I see it in Windows 10, so the "only" bit isn't true.

> we can get by without `Applications` or `AppPaths`.

Wait, do you mean for pre-XP installs? I thought the plan was to use `HKCR/Applications`? You can definitely stop using `AppPaths` though, as that only seems to get one install (the `chrome.exe` issue you mentioned).

**Comment by vweevers** [comment]
> No, it's not an executable name

From the [article](https://msdn.microsoft.com/en-us/library/windows/desktop/ee872121(v=vs.85).aspx) you linked to:

> Through the inclusion of registry entries under the **HKEY_CLASSES_ROOT\Applications\ApplicationName.exe** subkey, applications can provide the application-specific information shown in the following table.

To be clear, my plan was to read:

```
[HKEY_CLASSES_ROOT\.html\OpenWithProgids]
"ChromeHTML"=""
"ChromeSSHTM.FE4TJPY7L34RHQC5HHQTIPH5JM"=""
"ChromiumHTM.FE4TJPY7L34RHQC5HHQTIPH5JM"=""
```

Then for each of those, taking `ChromeHTML` as an example, read the values under `HKCR\ChromeHTML\Application` as well as `HKCR\ChromeHTML\shell\open\command`. That should give us a name and executable path.

I don't want to change the detection method for other browsers just yet. Things we learn here can be useful in the future, but let's focus on Chrome first.

**Comment by vweevers** [comment]
BTW, here's the [code from Chromium itself](https://github.com/nwjs/chromium.src/blob/85293a87a64f32a5679734ce7b645699142bd60a/chrome/installer/util/google_update_settings.cc#L141) to determine the channel name (as displayed on `chrome://help/`). A crazy thought is to spawn Chrome, attach a debugger, open a tab to `chrome://help-frame/strings.js` and read `browserVersion`. A saner thought is that the `HKCU\SOFTWARE\Google\Update\ClientState\[guid]\ap` might be useful after all, for multi-installs at least, because apparently that guid is hard-coded in chromium.

**Comment by tomitm** [comment]
Yeah, I'm guessing maybe it was recommended to use the executable name originally, then due to conflicts, people that wasn't the greatest idea? Considering how that key isn't actually used to execute, rather the value of `Shell\open\command` is. For example, it looks like Windows Store apps use `AppX{id}` anyway.

Totally on board with your plan, except the `FE4TJPY...` value is unique to each machine. For example, mine is `ZG6HR7E...`. Is it possible to match using `/^ChromeSSHTML.*/`?

Other than that, that sounds like it would work great. :+1:

**Comment by vweevers** [comment]
Yes, it'll probably be something like:

```js
const names = ['ChromeHTML', 'ChromeSSHTM', 'ChromiumHTM']
values.filter(v => names.indexOf(v.split('.')[0]) >= 0)
```

**Comment by tomitm** [comment]
Isn't that Chromium code part of the installer? Wonder where `strings.js` gets populated from.

**Comment by vweevers** [comment]
> Isn't that Chromium code part of the installer? Wonder where strings.js gets populated from.

I tracked it from [here](https://github.com/nwjs/chromium.src/blob/85293a87a64f32a5679734ce7b645699142bd60a/chrome/browser/ui/webui/help/help_handler.cc#L300) to [here](https://github.com/nwjs/chromium.src/blob/85293a87a64f32a5679734ce7b645699142bd60a/chrome/browser/ui/webui/help/help_handler.cc#L405) to [here](https://github.com/nwjs/chromium.src/blob/1088cfc266e0f8a533fa1d22001c4c5f856245d5/chrome/common/channel_info_win.cc#L20) to [here](https://github.com/nwjs/chromium.src/blob/85293a87a64f32a5679734ce7b645699142bd60a/chrome/installer/util/google_update_settings.cc#L469) ;)

**Comment by tomitm** [comment]
Found [something even closer](https://github.com/nwjs/chromium.src/blob/85293a87a64f32a5679734ce7b645699142bd60a/components/version_info/version_info.cc) :wink:

[Generated](https://github.com/nwjs/chromium.src/blob/85293a87a64f32a5679734ce7b645699142bd60a/components/version_info/version_info_values.h.version) at build time, I guess.

**Comment by tomitm** [comment]
Came across [this bug report](https://bugs.chromium.org/p/chromium/issues/detail?id=158372), which basically states we'll never be able to use `chrome --version` from the command line, bugger.

**Comment by vweevers** [comment]
> Generated at build time, I guess.

I wish it was, because then we would be able to extract it (by reading the binary) (even better would be if they made the channel name part of the executable's `ProductName`). For Chromium, the channel name is hardcoded. For Chrome it isn't as far I can tell. Because of the whole multi-install mess they need that `GetChromeChannelInternal` function I first linked to.

ShellIntegration has some interesting things though. Lol: [`ShellIntegration::IsFirefoxDefaultBrowser()`](https://github.com/nwjs/chromium.src/blob/6e883e40d456dd422ac08ced1c25e7403f78bada/chrome/browser/shell_integration_win.cc#L372)

> we'll never be able to use chrome --version from the command line. Plus the fact that they've removed --version for Windows.

Yeah, it's sad. Firefox has a similar issue, you have to redirect the output like `firefox --version | more`. I did exactly that in an earlier version of `win-detect-browsers`, but it was too slow

**Comment by vweevers** [comment]
Ah, that commit you linked to clarifies things

> Create a new component version_info that allows to get information about
the current version of Chrome running, except for the channel (which has
to be provided by the embedder).
>
>The method to get the channel is not moved to the component because on
Windows it depends on //chrome/installer which would be non-trivial to
componentize. Since the method is embedder-specific (though it have no
bad dependencies on the other platform) it makes sense to not move it
into the component.

**Comment by vweevers** [comment]
> A crazy thought is to spawn a hidden Chrome, attach a debugger, open a tab to chrome://help-frame/strings.js and read browserVersion.

Uh, this actually works. Depends on some private code so I can't share it just yet, but I did not expect that

**Comment by tomitm** [comment]
So, is it just me, or does that commit message imply that the installer is the root source of the channel info? In which case, `HKCU\SOFTWARE\Google\Update\ClientState\{guid}\ap`?

> Uh, this actually works. Depends on some private code so I can't share it just yet, but I did not expect that
Awesome! It **does** sound crazy, but do you think it makes sense as a viable solution? You could probably extend that to other browsers as well... How fast can that method get a channel?

**Comment by vweevers** [comment]
> So, is it just me, or does that commit message imply that the installer is the root source of the channel info?

I'm not entirely sure what you mean, when you say **_the_** installer. But I think what the commit author meant was everything in [/chrome/installer](https://github.com/nwjs/chromium.src/blob/85293a87a64f32a5679734ce7b645699142bd60a/chrome/installer), like `chrome/installer/util/google_update_settings.cc`, which defines the `GetChromeChannelInternal` function. There's no real single source of truth, but I consider this function to be the heart of what we're talking about. It's included in [various places](https://github.com/nwjs/chromium.src/search?utf8=%E2%9C%93&q=google_update_settings.c) and ends up compiled in `chrome.exe` itself.

> Awesome! It does sound crazy, but do you think it makes sense as a viable solution? You could probably extend that to other browsers as well... How fast can that method get a channel?

It's a cool hack, and satisfied my curiosity, but it takes ~950 ms. That's 5x the time spent by `win-detect-browsers` to find all 13 browsers on my system. Even if I optimized the code and disabled a lot of Chrome features (e.g. `--disable-extensions`), it's still a process to spawn, and Chrome has to build a profile, start the debugger, et cetera. And there are other issues:

- The Remote Debugging Protocol is not that stable
- Location and format of `chrome://help-frame` neither
- Unit testing and debugging could be a pain

So.. nah, not viable. `OpenWithProgids` is still our best bet.

**Comment by tomitm** [comment]
Oh, sorry! I'm not completely sure what "_the installer_" is either. Definitely `/chrome/installer`, but what is that actually? My guess was Google Updater, as in the service that lives in the background and keeps Chrome updated, but I feel like I might be wrong on that.

I see a bunch of [references to `ClientState`](https://github.com/nwjs/chromium.src/search?utf8=%E2%9C%93&q=ClientState) for example (one of the `HKCU\SOFTWARE\Google\Update` keys) under `/chrome/installer`, hence why I think it's the updater.

So if it is the updater, that leads me to think that using those registry values as the source for the channel should be suitable. As in `HKCU\SOFTWARE\Google\Update\ClientState\{guid}\ap`.

> It's a cool hack, and satisfied my curiosity, but it takes ~950 ms.

Yikes, that's a bit much. Still pretty neat though!

**Comment by vweevers** [comment]
What the main purpose of `/chrome/installer` is, is not a relevant question. There is no single component or service that has the responsibility of providing channel information. That's the problem. Because of the situation on Windows, where installation state is less straight-forward, the `chrome://help` page (one of very few places where channel info is displayed) has to borrow a bunch of functions from `/chrome/installer`.

But yes, like I said before, the `ClientState` keys can be useful for some but not all cases.

I'll get back to this in a few days

**Comment by vweevers** [comment]
@tomitm this has been implemented, with both the `ClientState` and `Progids` methods. Will be a while before it's released (see #28). The output looks like this:

```
{ name: 'chrome',
  version: '57.0.2931.0',
  path: 'C:\\Users\\vweevers\\AppData\\Local\\Google\\Chrome SxS\\Application\\chrome.exe',
  channel: 'canary',
  bitness: 64,
  guid: '4EA16AC7-FD5A-47C3-875B-DBF4A2008C20',
  info:
   { FileVersion: '57.0.2931.0',
     CompanyName: 'Google Inc.',
     FileDescription: 'Google Chrome',
     InternalName: 'chrome_exe',
     LegalCopyright: 'Copyright 2016 Google Inc. All rights reserved.',
     OriginalFilename: 'chrome.exe',
     ProductName: 'Google Chrome',
     ProductVersion: '57.0.2931.0',
     CompanyShortName: 'Google',
     ProductShortName: 'Chrome',
     LastChange: '5059e49d0a21c1c311653c1613de16b27aefc963-refs/heads/master@{#434310}',
     'Official Build': '1' },
  uninstall:
   { path: 'C:\\Users\\vweevers\\AppData\\Local\\Google\\Chrome SxS\\Application\\57.0.2931.0\\Installer\\setup.exe',
     arguments: [ '--uninstall', '--chrome-sxs' ] } }
{ name: 'chrome',
  version: '54.0.2840.99',
  path: 'C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe',
  channel: 'stable',
  bitness: 32,
  guid: '8A69D345-D564-463C-AFF1-A69D9E530F96',
  info:
   { FileVersion: '54.0.2840.99',
     CompanyName: 'Google Inc.',
     FileDescription: 'Google Chrome',
     InternalName: 'chrome_exe',
     LegalCopyright: 'Copyright 2016 Google Inc. All rights reserved.',
     OriginalFilename: 'chrome.exe',
     ProductName: 'Google Chrome',
     ProductVersion: '54.0.2840.99',
     CompanyShortName: 'Google',
     ProductShortName: 'Chrome',
     LastChange: '7eca4ce1e662f12cadaf96c30cd2335fd03e7140-refs/branch-heads/2840@{#830}',
     'Official Build': '1' },
  uninstall:
   { path: 'C:\\Program Files (x86)\\Google\\Chrome\\Application\\54.0.2840.99\\Installer\\setup.exe',

     arguments: [ '--uninstall', '--multi-install', '--chrome', '--system-level' ] } }

Found 2 browsers in 11 ways within 211ms.
```
